File: 백준/Silver/1992. 쿼드트리/쿼드트리.java
import java.io.*;

public class Main {
    private static char[][] arr;
    private static BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));


    public static void div(int i, int j,int size) throws IOException {
        if (check(i, j,size)) {
            bw.write(arr[i][j]);
            return;
        } else {
            bw.write("(");
            size = size / 2;
            div(i, j, size);
            div(i, j+size, size);
            div(i+size, j, size);
            div(i+size, j+size, size);
        }
        bw.write(")");
    }

    static boolean check(int i, int j,int size) {
        int save = arr[i][j];
        for (int k = 0; k < size; k++) {
            for (int l = 0; l < size; l++) {
                if ( save != arr[i+k][j+l]) {
                    return false;
                }
            }
        }
        return true;
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int N = Integer.parseInt((br.readLine()));

        arr = new char[N][N];

        for (int i = 0; i < arr.length; i++) {
            String str = br.readLine();
            for (int j = 0; j < str.length(); j++) {
                arr[i][j] = str.charAt(j);
            }
        }

        div(0, 0,N);

        bw.flush();
        bw.close();
    }
}

File: 백준/Silver/1992. 쿼드트리/쿼드트리.cc
#include<iostream>
using namespace std;

int arr[64][64];

bool check(int y, int x, int size) {
	int first = arr[y][x];
	for (int i = y; i < y + size; i++) {
		for (int j = x; j < x + size; j++) {
			if (arr[i][j] != first) {
				cout << "(";
				return false;
			}
		}
	}
	return true;
}

void div(int y, int x, int size) {
	if (size == 1) {
		cout << arr[y][x];
		return;
	}

	if (!check(y, x, size)) {
		for (int i = 0; i < 2; i++) {
			for (int j = 0; j < 2; j++) {
				div(y + i * size / 2, x + j * size / 2, size / 2); // 4등분하기
			}
		}
		// 4등분 다 했으면 괄호 닫기
		cout << ")";
	}
	else {
		cout << arr[y][x];
		return;
	}

}

int main() {
	int n;
	cin >> n;

	for (int i = 0; i < n; i++) {
		string line;
		cin >> line; // 한 줄 전체 읽어오기
		for (int j = 0; j < n; j++) {
			arr[i][j] = line[j] - '0'; // 숫자로 바꾸기
		}
	}

	div(0, 0, n);
}